### PR TITLE
Phase2-hgx233 Correct the HGCal code for working with dd4hep

### PR DESCRIPTION
**Pull request by bsunanda**
#### PR description:

Correct the interface of Reco geometry of HGCal for dd4hep interface

#### PR validation:

Use standard runTheMatrix.py for these tests

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

Nothing special

### CODE REVIEW

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by cvuosalo** [comment]
@bsunanda Please add to the PR description whether this PR completes the Reco geometry migration for HGCal. If it does not, please explain in the description how much still needs to be done. Thanks.

**Comment by cvuosalo** [comment]
+1

**Comment by bsunanda** [comment]
@kpedro88 Can you sign off this

**Comment by bsunanda** [comment]
@cvuosalo I am working on the transition of test beam files. Hope to solve that soon, but I need this PR to be integrated first

File: Geometry/HGCalCommonData/src/HGCalGeomParameters.cc [hunk 1]
```diff
@@ -688,118 +678,142 @@ void HGCalGeomParameters::loadGeometryHexagon8(const DDFilteredView& _fv, HGCalP
...
+      int zside = (nsiz > php.levelZSide_) ? copy[php.levelZSide_] : -1;
+      if (zside != 1)
+        zside = -1;
+      const DDSolid& sol = fv.logicalPart().solid();
+#ifdef EDM_ML_DEBUG
+      edm::LogVerbatim("HGCalGeom") << sol.name() << " shape " << sol.shape() << " size " << nsiz << ":" << levelTop
+                                    << " lay " << lay << " z " << zside;
+#endif
+      if (lay == 0) {
+        edm::LogError("HGCalGeom") << "Funny layer # " << lay << " zp " << zside << " in " << nsiz << " components";
```
Comment (kpedro88) [inline Geometry/HGCalCommonData/src/HGCalGeomParameters.cc]: no reason to emit LogError right before throwing an exception, just combine the two (the LogError message is more informative, so it should be used in the exception)

File: Geometry/HGCalCommonData/src/HGCalGeomParameters.cc [hunk 2]
```diff
@@ -1691,3 +1704,9 @@ std::pair<double, double> HGCalGeomParameters::cellPosition(
 void HGCalGeomParameters::rescale(std::vector<double>& v, const double s) {
   std::for_each(v.begin(), v.end(), [s](double& n) { n *= s; });
 }
+
+void HGCalGeomParameters::resetZero(std::vector<double>& v) {
+  for (auto n : v)
+    if (std::abs(n) < 1.e-20)
```
Comment (kpedro88) [inline Geometry/HGCalCommonData/src/HGCalGeomParameters.cc]: magic number `1.e-20` should be named constant

File: Geometry/HGCalCommonData/src/HGCalGeomParameters.cc (hunk 1 above)
Comment (bsunanda) [reply Geometry/HGCalCommonData/src/HGCalGeomParameters.cc]: Done

File: Geometry/HGCalCommonData/src/HGCalGeomParameters.cc (hunk 2 above)
Comment (bsunanda) [reply Geometry/HGCalCommonData/src/HGCalGeomParameters.cc]: Done

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by kpedro88** [comment]
+upgrade

**Comment by bsunanda** [comment]
@cvuosalo Could you please sign this again? I am waiting it to be integrated before I can have the next PR

**Comment by cvuosalo** [comment]
@bsunanda Could you please add to the PR description how much of the Reco geometry migration for HGCal this PR completes? Please give an idea in the description how much still needs to be done. Thanks.

**Comment by cvuosalo** [comment]
+1

**Comment by silviodonato** [comment]
+1

**Pull request merged**